packages/controller-utils/src/create-service-policy.test.ts

```diff
@@ -0,0 +1,2703 @@
+                // ignore any errors - adding to the promise queue is enough
+              });
+              await promise;
+
+              expect(onDegraded).toHaveBeenCalledTimes(1);
+            });
+          });
+
+          describe('using a custom degraded threshold', () => {
+            it('does not call the onDegraded callback if the service execution time is below the threshold', async () => {
+              const degradedThreshold = 2000;
+              const maxConsecutiveFailures = DEFAULT_MAX_RETRIES + 1;
+              let invocationCounter = 0;
+              const error = new Error('failure');
+              const mockService = () => {
+                invocationCounter += 1;
+                if (invocationCounter === DEFAULT_MAX_RETRIES + 1) {
+                  return { some: 'data' };
+                }
+                throw error;
+              };
+              const onDegraded = jest.fn();
+              const policy = createServicePolicy({
+                maxConsecutiveFailures,
+                onDegraded,
+                degradedThreshold,
+              });
+
+              const promise = policy.execute(() => mockService());
+              clock.runAllAsync().catch(() => {
+                // ignore any errors - adding to the promise queue is enough
+              });
+              await promise;
+
+              expect(onDegraded).not.toHaveBeenCalled();
+            });
+
+            it('calls the onDegraded callback once if the service execution time is beyond the threshold', async () => {
+              const degradedThreshold = 2000;
+              const maxConsecutiveFailures = DEFAULT_MAX_RETRIES + 1;
+              const delay = degradedThreshold + 1;
+              let invocationCounter = 0;
+              const mockService = () => {
+                invocationCounter += 1;
+                return new Promise((resolve, reject) => {
+                  if (invocationCounter === DEFAULT_MAX_RETRIES + 1) {
+                    setTimeout(() => resolve({ some: 'data' }), delay);
+                  } else {
+                    reject(new Error('failure'));
+                  }
+                });
+              };
+              const onDegraded = jest.fn();
+              const policy = createServicePolicy({
+                maxConsecutiveFailures,
+                onDegraded,
+                degradedThreshold,
+              });
+
+              const promise = policy.execute(() => mockService());
+              clock.runAllAsync().catch(() => {
+                // ignore any errors - adding to the promise queue is enough
+              });
+              await promise;
+
+              expect(onDegraded).toHaveBeenCalledTimes(1);
+            });
+          });
+        });
+
+        describe('if the initial run + retries is greater than the max number of consecutive failures', () => {
+          it('throws a BrokenCircuitError before the service can succeed', async () => {
```

Comment:
This is a bit buried, but this (and other tests like it) is where we ensure that once the circuit is open, and before the circuit break duration has elapsed, any subsequent executions are essentially refused.